### PR TITLE
fuzz: makes cve 2021 3449 reachable by fuzzer

### PR DESCRIPTION
**Pull request by catenacyber**
##### Checklist
- [x] tests are added or updated

Improves fuzzing efficiency by adding `#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION ` blocks to allow fuzzer not to have the right hmac  and such...

Base64 reproducer is then `FgMBAJ4BAACaAwN2WCKGUw4FEFIYi3F907Sb5MDmWelWy29RN8iXGeK2MCCQpnjzvNz8NwRwpRlsWBHBwo+InPaUtgDZA2iX0W3ZRAACAAIBAABPAAUABQEAAAAAAAoACgAIAB0AFwAYABkACwACAQAADQAaABgIBAQDCAcIBQgGBAEFAQYBBQMGAwIBAgP/AQABAAASAAAAKwAHBgMDAwIDARYDAwEGEAABAgEAKIhLewbvjA5ZKexgA5M/4aWtnofAEa6bYqSX3nkxQVjHobGYGcFVEdjkw8AuLv73wCpLYAUb9Lji67lNdR4V/A5CI6Ji5c2Vz6Gt3Ic33xu94UxjeDJ4bx1s0rL26sja53pM7GTFHJjxN0JkrCv2tVeLmeTFptxIml5OY/HW6JNBb0b7xtpkr+KD3SS86SXLpFj+yac0usy22kpQ6AWWXH/gPu+h9NEIHj8hwTNUsuBVkunkSb4dP3mMjc9WEZXZ1uob3RC9+QPq/+NVVUaeUtlpiW6Ps1OwLCgfx3OwUVnmE9Y33G8ypJ2SPxq8Hw83C9pvj8DuKJbXuNTOnLC2iRQDAwABARYDAwAkFAAADI1vNle2WP9to22VSwAAAAAAAAAAAAAAAAAAAAAAAAAAFgMDAL4BAACmAwNzQHup5P7r66jCKYDvTWMwDt5S+ZH4PB3bB2MrM/LGrSD7KrZcwEPQ60fpS+cYSMSXAIbep9ED5i0oqwN6TmAg6QACAAIBAABbAAUABQEAAAAAAAoACgAIAB0AFwAYABkACwACAQAAMgAaABgIBAQDCAcIBQgGBAEFAQYBBQMGAwIBAgP/AQANDI1vNle2WP9to22VSwASAAAAKwAHBgMDAwIDAQAAAAAAAAAAAAAAAAAAAAAAAAAAFQMDABYBAAAAAAAAAAAAAAAAAAAAAAAAAAAA`

### CODE REVIEW

File: ssl/record/methods/tls_common.c [hunk 1]
```diff
@@ -863,6 +863,11 @@ int tls_get_more_records(OSSL_RECORD_LAYER *rl)
                 enc_err = 0;
             if (thisrr->length > SSL3_RT_MAX_COMPRESSED_LENGTH + mac_size)
                 enc_err = 0;
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+            if (enc_err == 0 && mac_size > 0 && (md[0] ^ thismb->mac[0]) != 0xFF) {
```
Comment (catenacyber) [inline ssl/record/methods/tls_common.c:867]: I try to have the fuzzer still get errors, but only if every bit of the first byte is wrong ie `(md[0] ^ thismb->mac[0]) == 0xFF`

File: ssl/ssl_sess.c [hunk 2]
```diff
@@ -298,10 +298,15 @@ static int def_generate_session_id(SSL *ssl, unsigned char *id,
                                    unsigned int *id_len)
 {
     unsigned int retry = 0;
-    do
+    do {
         if (RAND_bytes_ex(ssl->ctx->libctx, id, *id_len, 0) <= 0)
             return 0;
-    while (SSL_has_matching_session_id(ssl, id, *id_len) &&
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
```
Comment (catenacyber) [inline ssl/ssl_sess.c:304]: No cryptography here, but `RAND_bytes_ex` gives always the same result when built for fuzzing, so this function ends failing, unless we increment the result at each try...

**Comment by catenacyber** [comment]
I have already committed to openssl, did I not sign the CLA ?

**Comment by catenacyber** [comment]
For information, I have not found new bugs with some hours of fuzzing server with this...

**Comment by paulidale** [comment]
> I have already committed to openssl, did I not sign the CLA ?

The email address being used for these commits is <contact@catenacyber.fr> which isn't the address we've got on file.

**Comment by catenacyber** [comment]
> The email address being used for this commit is [contact@catenacyber.fr](mailto:contact@catenacyber.fr) which isn't the address we've got on file. Changing the address to what was registers and force pushing will fix this.

Indeed, used the right git config and force-pushed, thanks Paul

cf https://github.com/openssl/openssl/commits?author=catenacyber

**Comment by catenacyber** [comment]
For info, I realize this is a lot like boringssl commit 9343b0b8b3

**Review by paulidale — APPROVED** [review]

File: ssl/statem/extensions_srvr.c [hunk 3]
```diff
@@ -58,8 +58,13 @@ int tls_parse_ctos_renegotiate(SSL_CONNECTION *s, PACKET *pkt,
...
     }
 
+#ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
     if (memcmp(data, s->s3.previous_client_finished,
-               s->s3.previous_client_finished_len)) {
+               s->s3.previous_client_finished_len))
+#else
+    if (s->s3.previous_client_finished_len > 0 && (data[0] ^ s->s3.previous_client_finished[0]) == 0xFF)
+#endif
+    {
```
Comment (kroeckx) [inline ssl/statem/extensions_srvr.c:67]: I think the { belongs on the same line as the if, and the new line is too long.

File: ssl/record/methods/tls_common.c (hunk 1 above)
Comment (kroeckx) [reply ssl/record/methods/tls_common.c:867]: Any reason you pick 0xFF and not say 0x0? It doesn't really matter for me.

**Review by kroeckx — APPROVED** [review]
Assuming the style issues get resolved.

**Comment by kroeckx** [comment]
There are probably other improvements by adding FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION that can be done. For instance using an encryption that just returns the plain text should make it a lot easier for the fuzzer. You can also do the same thing for the client fuzzer.

**Comment by catenacyber** [comment]
> There are probably other improvements by adding FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION that can be done. 

Indeed, there are probably more improvements to be made (cf other calls to `CRYPTO_memcmp`)...
@davidben do you have advices about it ?

I found those by trying to have CVE 2021 3449 reachable by fuzzer.
Once I got it, I stopped.

> For instance using an encryption that just returns the plain text should make it a lot easier for the fuzzer. You can also do the same thing for the client fuzzer.

There is already encryption that does plain text (but there is still a HMAC cf change in tls_common.c).
I use `TLS_RSA_WITH_NULL_SHA` to reproduce the vulnerability.

**Comment by catenacyber** [comment]
Do you want me to squash the commits together ?

**Comment by paulidale** [comment]
We usually squash when merging.
Feel free to if there is anything specific or complicated you'd like there.

**Comment by catenacyber** [comment]
> We usually squash when merging.
> Feel free to if there is anything specific or complicated you'd like there.

Nothing specific or complicated : I would just keep the first commit message

**Comment by t8m** [comment]
Also if you want to make it obvious that some commits should be squashed when merging, you can use `git commit --fixup`

**Comment by t8m** [comment]
Merged to master branch. Thank you for your contribution.

**Comment by paulidale** [comment]
Coverity is reporting problems with this:
```c
New defect(s) Reported-by: Coverity Scan
Showing 3 of 3 defect(s)


** CID 1520807:  Incorrect expression  (PRECEDENCE_ERROR)
/ssl/statem/statem_lib.c: 839 in tls_process_finished()


________________________________________________________________________________________________________
*** CID 1520807:  Incorrect expression  (PRECEDENCE_ERROR)
/ssl/statem/statem_lib.c: 839 in tls_process_finished()
833         }
834     
835         ok = CRYPTO_memcmp(PACKET_data(pkt), s->s3.tmp.peer_finish_md,
836                            md_len);
837     #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
838         if (ok != 0) {
>>>     CID 1520807:  Incorrect expression  (PRECEDENCE_ERROR)
>>>     Evaluates as: "PACKET_data(pkt)[0] ^ (s->s3.tmp.peer_finish_md[0] != 255)", because "!=" has higher operator precedence than "^". The context suggests that this might be in error.
839             if (PACKET_data(pkt)[0] ^ s->s3.tmp.peer_finish_md[0] != 0xFF) {
840                 ok = 0;
841             }
842         }
843     #endif
844         if (ok != 0) {

** CID 1520806:  Null pointer dereferences  (FORWARD_NULL)
/ssl/record/methods/tls_common.c: 868 in tls_get_more_records()


________________________________________________________________________________________________________
*** CID 1520806:  Null pointer dereferences  (FORWARD_NULL)
/ssl/record/methods/tls_common.c: 868 in tls_get_more_records()
862                 if (i == 0 || thismb == NULL || thismb->mac == NULL
863                     || CRYPTO_memcmp(md, thismb->mac, (size_t)mac_size) != 0)
864                     enc_err = 0;
865                 if (thisrr->length > SSL3_RT_MAX_COMPRESSED_LENGTH + mac_size)
866                     enc_err = 0;
867     #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
>>>     CID 1520806:  Null pointer dereferences  (FORWARD_NULL)
>>>     Dereferencing null pointer "thismb->mac".
868                 if (enc_err == 0 && mac_size > 0 && (md[0] ^ thismb->mac[0]) != 0xFF) {
869                     enc_err = 1;
870                 }
871     #endif
872             }
873         }

** CID 1520805:  Incorrect expression  (PRECEDENCE_ERROR)
/ssl/statem/extensions_srvr.c: 66 in tls_parse_ctos_renegotiate()


________________________________________________________________________________________________________
*** CID 1520805:  Incorrect expression  (PRECEDENCE_ERROR)
/ssl/statem/extensions_srvr.c: 66 in tls_parse_ctos_renegotiate()
60         }
61     
62         ok = memcmp(data, s->s3.previous_client_finished,
63                         s->s3.previous_client_finished_len);
64     #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
65         if (ok) {
>>>     CID 1520805:  Incorrect expression  (PRECEDENCE_ERROR)
>>>     Evaluates as: "data[0] ^ (s->s3.previous_client_finished[0] != 255)", because "!=" has higher operator precedence than "^". The context suggests that this might be in error.
66             if (data[0] ^ s->s3.previous_client_finished[0] != 0xFF) {
67                 ok = 0;
68             }
69         }
70     #endif
71         if (ok) {
```

**Comment by catenacyber** [comment]
Fixes proposed in https://github.com/openssl/openssl/pull/20253

**Comment by paulidale** [comment]
Thanks for the fix.

**Comment by catenacyber** [comment]
It is already helping fuzzing cf https://storage.googleapis.com/oss-fuzz-coverage/openssl/reports/20230208/linux/src/openssl/ssl/statem/statem_lib.c.html vs https://storage.googleapis.com/oss-fuzz-coverage/openssl/reports/20230209/linux/src/openssl/ssl/statem/statem_lib.c.html

**Pull request closed without merging**